packages/authorization-updater/src/index.ts

```diff
@@ -119,31 +176,139 @@ function processMessage(
+          {
+            event_version: 2,
+            type: "PurposeVersionSuspendedByConsumer",
+          },
+          {
+            event_version: 2,
+            type: "PurposeVersionSuspendedByProducer",
+          },
+          {
+            event_version: 2,
+            type: "PurposeVersionUnsuspendedByConsumer",
+          },
+          {
+            event_version: 2,
+            type: "PurposeVersionUnsuspendedByProducer",
+          },
+          {
+            event_version: 2,
+            type: "PurposeVersionOverQuotaUnsuspended",
+          },
+          {
+            event_version: 2,
+            type: "NewPurposeVersionActivated",
+          },
+          {
+            event_version: 2,
+            type: "NewPurposeVersionWaitingForApproval",
+          },
+          {
+            event_version: 2,
+            type: "PurposeVersionRejected",
+          },
+          {
+            event_version: 2,
+            type: "PurposeVersionActivated",
+          },
+          {
+            event_version: 2,
+            type: "PurposeArchived",
+          },
+          async (msg): Promise<void> => {
+            const { purposeId, purposeVersion } = getPurposeVersionFromEvent(
+              msg,
+              msg.type
+            );
+
+            await executeUpdate(msg.type, messagePayload, () =>
+              authService.updatePurposeState(
+                purposeId,
+                purposeVersion.id,
+                purposeVersion.state === purposeVersionState.active
+                  ? "ACTIVE"
+                  : "INACTIVE"
+              )
+            );
           }
         )
-        .otherwise(() => undefined);
-
-      if (updateSeed) {
-        await executeUpdate(updateSeed.eventType, messagePayload, () =>
-          authService.updateEServiceState(
-            ApiClientComponent.parse(updateSeed.state),
-            updateSeed.descriptorId,
-            updateSeed.eserviceId,
-            updateSeed.audience,
-            updateSeed.voucherLifespan
-          )
-        );
-      }
+        .with(
+          {
+            event_version: 2,
+            type: "PurposeActivated",
+          },
+          {
+            event_version: 2,
+            type: "PurposeWaitingForApproval",
+          },
+          async (msg): Promise<void> => {
+            const purpose = getPurposeFromEvent(msg, msg.type);
+
+            const purposeVersion = purpose.versions[0];
+
+            if (!purposeVersion) {
+              throw missingKafkaMessageDataError("purposeVersion", msg.type);
+            }
+
+            await executeUpdate(msg.type, messagePayload, () =>
+              authService.updatePurposeState(
+                purpose.id,
+                purposeVersion.id,
+                purposeVersion.state === purposeVersionState.active
+                  ? "ACTIVE"
+                  : "INACTIVE"
+              )
+            );
+          }
+        )
+        .otherwise(() => void 0);
```

Comment:
In the previous implementation there was no handling regarding unknown incoming messages. 
Should we add a log/throw in case no message matches?

Reply:
at first I was of the same opinion.
Point is that from now on we will create several consumers and it may be easy to forget one of them when creating a new event.

What if we use an exhaustive match (maybe split by domain), so that the compiler would help us here?
In this case when adding a new event the compiler will already remind us that we need to handle it, even if it is just a no-op.
Obviously we should handle only messages from domains interested in the consumer.

What do you think?

Reply:
I agree with @galales, his proposal gives us some advantages in terms of bug prevention and code organization.